vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/WidgetItemFactory.kt

```diff
@@ -0,0 +1,126 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.app.features.home.room.detail.timeline.factory
+
+import im.vector.app.ActiveSessionDataSource
+import im.vector.app.R
+import im.vector.app.core.epoxy.VectorEpoxyModel
+import im.vector.app.core.resources.StringProvider
+import im.vector.app.features.home.room.detail.timeline.TimelineEventController
+import im.vector.app.features.home.room.detail.timeline.helper.AvatarSizeProvider
+import im.vector.app.features.home.room.detail.timeline.helper.MessageInformationDataFactory
+import im.vector.app.features.home.room.detail.timeline.helper.MessageItemAttributesFactory
+import im.vector.app.features.home.room.detail.timeline.helper.RoomSummaryHolder
+import im.vector.app.features.home.room.detail.timeline.item.WidgetTileTimelineItem
+import im.vector.app.features.home.room.detail.timeline.item.WidgetTileTimelineItem_
+import org.matrix.android.sdk.api.extensions.orFalse
+import org.matrix.android.sdk.api.session.events.model.Event
+import org.matrix.android.sdk.api.session.events.model.toModel
+import org.matrix.android.sdk.api.session.room.timeline.TimelineEvent
+import org.matrix.android.sdk.api.session.widgets.model.WidgetContent
+import org.matrix.android.sdk.api.session.widgets.model.WidgetType
+import javax.inject.Inject
+
+class WidgetItemFactory @Inject constructor(
+        private val sp: StringProvider,
+        private val roomSummaryHolder: RoomSummaryHolder,
+        private val messageItemAttributesFactory: MessageItemAttributesFactory,
+        private val informationDataFactory: MessageInformationDataFactory,
+        private val noticeItemFactory: NoticeItemFactory,
+        private val avatarSizeProvider: AvatarSizeProvider,
+        private val activeSessionDataSource: ActiveSessionDataSource
+) {
+    private val currentUserId: String?
+        get() = activeSessionDataSource.currentValue?.orNull()?.myUserId
+
+    private fun Event.isSentByCurrentUser() = senderId != null && senderId == currentUserId
+
+    fun create(event: TimelineEvent,
+               highlight: Boolean,
+               callback: TimelineEventController.Callback?): VectorEpoxyModel<*>? {
+        val widgetContent: WidgetContent = event.root.getClearContent().toModel() ?: return null
+        val previousWidgetContent: WidgetContent? = event.root.resolvedPrevContent().toModel()
+
+        return when (WidgetType.fromString(widgetContent.type ?: previousWidgetContent?.type ?: "")) {
+            WidgetType.Jitsi       -> createJitsiItem(event, callback, widgetContent, previousWidgetContent)
+            WidgetType.TradingView,
```

Comment:
use else instead?

Reply:
Yes, it was to show all the cases we could improve... I will change to `else`

Reply:
Done